File: app/src/main/java/net/itsjustsomedude/tokens/NotificationReader.java
package net.itsjustsomedude.tokens;

import android.app.Notification;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.provider.Settings;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;

import android.util.Log;
import android.widget.Toast;

import net.itsjustsomedude.tokens.db.CoopRepository;
import net.itsjustsomedude.tokens.db.Event;
import net.itsjustsomedude.tokens.db.EventRepository;

import java.util.Arrays;
import java.util.Calendar;
import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class NotificationReader {
	private static final String TAG = "NotificationReader";

	private static final List<String> ALLOWED_PACKAGES = Arrays.asList(
			"com.auxbrain.egginc",
			"net.itsjustsomedude.tokens",
			"net.itsjustsomedude.tokens.debug"
	);

	private static final Pattern personCoopRegex = Pattern.compile("^(.+) \\((.+)\\) has (?:sent you|hatched).+?$");
	private static final Pattern tokenCountRegex = Pattern.compile("(?<=has sent you a gift of )([0-9]+)");

	public static final String PREF_ENABLED = "service_control_enable_service";
	public static final String PREF_DISMISS = "auto_dismiss";
	// ^^^ Set in the root_preferences file.
	private static final int dismissDelay = 10 * 1000;

	public static void processAllNotifications() {
		NotificationService i = NotificationService.getInstance();
		if (i != null)
			i.processAllNotifications();
	}

	public static boolean isServiceEnabled(Context ctx) {
		return PreferenceManager.getDefaultSharedPreferences(ctx).getBoolean(PREF_ENABLED, true);
	}

	public static boolean isServiceRunning() {
		NotificationService i = NotificationService.getInstance();
		return i != null;
	}

	public static void setServiceEnabled(Context ctx, boolean enable) {
		PreferenceManager.getDefaultSharedPreferences(ctx)
				.edit()
				.putBoolean(PREF_ENABLED, enable)
				.apply();

		NotificationService service = NotificationService.getInstance();
		if (enable && service == null) {
			// Enable Service, it's not running.
			ComponentName cn = new ComponentName(ctx, NotificationService.class);
			String flat = Settings.Secure.getString(ctx.getContentResolver(), "enabled_notification_listeners");

			if (flat != null && flat.contains(cn.flattenToString())) {
				Log.i(TAG, "Asking it to start via requestRebind.");
				// TODO: Offer enable/disable from settings if android below N.
				if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
					NotificationListenerService.requestRebind(new ComponentName(ctx, NotificationService.class));
				}
			} else {
				Toast.makeText(ctx, ctx.getString(R.string.enable_service_toast), Toast.LENGTH_SHORT).show();
				ctx.startActivity(
						new Intent("android.settings.ACTION_NOTIFICATION_LISTENER_SETTINGS")
				);
			}
		} else if (!enable && service != null) {
			service.stop();
		}
//		else if (enable && service != null) {
//			// Enable, already running.
//		}
//		else if (!enable && service == null) {
//			// Disable, not running.
//		}
	}

	public static class NotificationService extends NotificationListenerService {
		private Handler dismissHandler;
		private static NotificationService instance;

		public static NotificationService getInstance() {
			return instance;
		}

		@Override
		public void onListenerConnected() {
			instance = this;

			if (isServiceEnabled(this)) {
				dismissHandler = new Handler(Looper.getMainLooper());
			} else {
				stop();
			}
		}

		@Override
		public void onListenerDisconnected() {
			super.onListenerDisconnected();
			destroy();
		}

		@Override
		public void onDestroy() {
			super.onDestroy();
			destroy();
		}

		public void stop() {
			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
				requestUnbind();
			} else {
				stopSelf();
			}
		}

		private void destroy() {
			instance = null;
			if (dismissHandler != null) {
				dismissHandler.removeCallbacksAndMessages(null);
				dismissHandler = null;
			}
		}

		@Override
		public void onNotificationPosted(StatusBarNotification sbn) {
			if (!isServiceEnabled(this) || !shouldProcessNotification(sbn)) {
				return;
			}


			processNotification(new EventRepository(getApplication()), sbn);
		}

		private boolean shouldProcessNotification(StatusBarNotification sbn) {
			if (!ALLOWED_PACKAGES.contains(sbn.getPackageName())) return false;

			Notification innerNote = sbn.getNotification();
			if (innerNote == null) return false;
			if ((innerNote.flags & Notification.FLAG_GROUP_SUMMARY) != 0) return false;

			CharSequence t = innerNote.extras.getCharSequence(Notification.EXTRA_TITLE);
			String title = t != null ?
					t.toString()
					: "";

			if (!title.contains("Gift Received")) return false;

			return true;
		}

		private void processNotification(EventRepository repository, StatusBarNotification sbn) {
			int id = sbn.getId();

			Notification innerNote = sbn.getNotification();

			String key = sbn.getKey();

			CharSequence bigText = innerNote.extras.getCharSequence(Notification.EXTRA_BIG_TEXT);
			CharSequence extraText = innerNote.extras.getCharSequence(Notification.EXTRA_TEXT);

			String text = bigText != null ?
					bigText.toString() :
					extraText != null ?
							extraText.toString() :
							"";

			// TODO: Add a pre-processor that extracts just the needed info.
			CharSequence t = innerNote.extras.getCharSequence(Notification.EXTRA_TITLE);
			String title = t != null ?
					t.toString()
					: "";

			String group = innerNote.getGroup();
			Calendar when = Calendar.getInstance();
			when.setTimeInMillis(innerNote.when);

			Log.i(TAG, "Processing note " + group + id + text);

			Matcher matches = personCoopRegex.matcher(text);
			if (!matches.lookingAt() || matches.groupCount() < 2) {
				Log.e(TAG, "Person/Coop Regex didn't match for notification content:");
				Log.e(TAG, text);
				return;
			}

			String person = matches.group(1);
			String coopName = matches.group(2);

			assert coopName != null;

			if (repository.blockingExists(coopName, group, id)) {
				Log.i(TAG, "Skipping notification that has already been processed.");
				scheduleRemoval(key);
				return;
			}

			if (title.contains("🐣")) {
				Log.i(TAG, "Processing CR");
				// Count of 0 indicates a CR.
				repository.blockingInsert(new Event(coopName, group, when, 0, person, Event.DIRECTION_SENT, id));
				scheduleRemoval(key);
				return;
			}

			int amount = 0;
			if (text.contains("has sent you a Boost Token"))
				amount = 1;
			else {
				try {
					Matcher countMatch = tokenCountRegex.matcher(text);
					if (!countMatch.find() || countMatch.group(1) == null) {
						Log.e(TAG, "Count Regex didn't match for notification content:");
						Log.e(TAG, text);
						return;
					}
					amount = Integer.parseInt(Objects.requireNonNull(countMatch.group(1)));
				} catch (NumberFormatException err) {
					Log.e(TAG, "Bad number in note:");
					Log.e(TAG, text);
				}
			}

			repository.blockingInsert(new Event(coopName, group, when, amount, person, Event.DIRECTION_SENT, id));
			Log.i(TAG, "Added event from note:");
			Log.i(TAG, text);

			scheduleRemoval(key);
		}

		public void processAllNotifications() {
			if (!isServiceEnabled(this)) {
				return;
			}

			EventRepository repo = new EventRepository(getApplication());

			StatusBarNotification[] activeNotifications = getActiveNotifications();
			for (StatusBarNotification sbn : activeNotifications) {
				if (shouldProcessNotification(sbn))
					processNotification(repo, sbn);
			}
		}

		private void scheduleRemoval(String key) {
			if (shouldDismiss())
				dismissHandler.postDelayed(() -> cancelNotification(key), dismissDelay);
		}

		// TODO: Make less calls to shared prefs, cache this.
		private boolean shouldDismiss() {
			return PreferenceManager.getDefaultSharedPreferences(this).getBoolean(PREF_DISMISS, false);
		}
	}
}

File: app/src/main/java/net/itsjustsomedude/tokens/NotificationReader.kt
package net.itsjustsomedude.tokens

import android.app.Notification
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.util.Log
import android.widget.Toast
import androidx.preference.PreferenceManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import net.itsjustsomedude.tokens.NotificationReader.NotificationService
import net.itsjustsomedude.tokens.db.Event
import net.itsjustsomedude.tokens.db.EventRepository
import org.koin.android.ext.android.getKoin
import java.util.Calendar
import java.util.regex.Pattern

object NotificationReader {
    const val PREF_ENABLED: String = "service_control_enable_service"
    const val PREF_DISMISS: String = "auto_dismiss"

    private const val TAG = "NotificationReader"

    private val personCoopRegex =
        Pattern.compile("^(.+) \\((.+)\\) has (?:sent you|hatched).+?$")
    private val tokenCountRegex = Pattern.compile("(?<=has sent you a gift of )([0-9]+)")

    // TODO: Set in the root_preferences file.
    private const val dismissDelay = 10 * 1000

    @JvmStatic
    fun processAllNotifications() {
        NotificationService.instance?.processAllNotifications()
    }

    @JvmStatic
    fun isServiceEnabled(ctx: Context): Boolean {
        return PreferenceManager.getDefaultSharedPreferences(ctx).getBoolean(PREF_ENABLED, true)
    }

    @JvmStatic
    val isServiceRunning: Boolean
        get() {
            val i = NotificationService.instance
            return i != null
        }

    @JvmStatic
    fun setServiceEnabled(ctx: Context, enable: Boolean) {
        PreferenceManager.getDefaultSharedPreferences(ctx)
            .edit()
            .putBoolean(PREF_ENABLED, enable)
            .apply()

        val service = NotificationService.instance
        if (enable && service == null) {
            // Enable Service, it's not running.
            val cn = ComponentName(ctx, NotificationService::class.java)
            val flat =
                Settings.Secure.getString(ctx.contentResolver, "enabled_notification_listeners")

            if (flat != null && flat.contains(cn.flattenToString())) {
                Log.i(TAG, "Asking it to start via requestRebind.")
                // TODO: Offer enable/disable from settings if android below N.
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    NotificationListenerService.requestRebind(
                        ComponentName(
                            ctx,
                            NotificationService::class.java
                        )
                    )
                }
            } else {
                Toast.makeText(
                    ctx,
                    ctx.getString(R.string.enable_service_toast),
                    Toast.LENGTH_LONG
                ).show()

                ctx.startActivity(
                    Intent("android.settings.ACTION_NOTIFICATION_LISTENER_SETTINGS")
                )
            }
        } else if (!enable && service != null) {
            service.stop()
        }
    }

    class NotificationService : NotificationListenerService() {
        private val serviceScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

        private var dismissHandler: Handler? = null

        override fun onListenerConnected() {
            instance = this

            if (isServiceEnabled(this)) {
                dismissHandler = Handler(Looper.getMainLooper())
            } else {
                stop()
            }
        }

        override fun onListenerDisconnected() {
            super.onListenerDisconnected()
            destroy()
        }

        override fun onDestroy() {
            super.onDestroy()
            destroy()
        }

        fun stop() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                requestUnbind()
            } else {
                // TODO: Open an old emulator and see if this works.
                stopSelf()
            }
        }

        private fun destroy() {
            serviceScope.cancel()

            instance = null

            if (dismissHandler != null) {
                dismissHandler!!.removeCallbacksAndMessages(null)
                dismissHandler = null
            }
        }

        fun processAllNotifications() {
            if (!isServiceEnabled(this))
                return

            val repo: EventRepository = getKoin().get()
            for (sbn in activeNotifications) {
                val sbnData = extractData(sbn)

                if (sbnData.shouldProcess)
                    serviceScope.launch {
                        processNotification(sbnData, repo)
                    }
            }
        }

        override fun onNotificationPosted(sbn: StatusBarNotification) {
            val note = extractData(sbn)

            if (!isServiceEnabled(this) || !note.shouldProcess)
                return

            serviceScope.launch {
                processNotification(note, getKoin().get())
            }
        }

        private suspend fun processNotification(
            note: ShortNotification,
            repository: EventRepository
        ) {
            val matches = personCoopRegex.matcher(note.text)
            if (!matches.lookingAt() || matches.groupCount() < 2) {
                Log.e(TAG, "Person/Coop Regex didn't match for notification content:")
                Log.e(TAG, note.text)
                return
            }

            val person = matches.group(1)!!
            val coopName = matches.group(2)!!

            if (repository.exists(coopName, note.group, note.id)) {
                Log.i(TAG, "Skipping notification that has already been processed.")
                scheduleRemoval(note.key)
                return
            }

            if (note.title.contains("🐣")) {
                Log.i(TAG, "Processing CR")
                repository.insert(
                    Event(
                        coop = coopName,
                        kevId = note.group,
                        person = person,
                        count = 0,      // Count of 0 indicates CR.
                        time = note.time,
                        direction = Event.DIRECTION_SENT,
                        notification = note.id
                    )
                )
                scheduleRemoval(note.key)
                return
            }

            val amount: Int
            if (note.text.contains("has sent you a Boost Token")) amount = 1
            else {
                try {
                    val countMatch = tokenCountRegex.matcher(note.text)

                    if (!countMatch.find() || countMatch.group(1) == null) {
                        Log.e(
                            TAG,
                            "Count Regex didn't match for notification content: ${note.text}"
                        )
                        return
                    }


                    amount = countMatch.group(1)!!.toInt()
                } catch (err: NumberFormatException) {
                    Log.e(TAG, "Bad number in note: ${note.text}")
                    return
                }
            }

            repository.insert(
                Event(
                    coop = coopName,
                    kevId = note.group,
                    person = person,
                    count = amount,
                    time = note.time,
                    direction = Event.DIRECTION_SENT,
                    notification = note.id
                )
            )

            Log.i(TAG, "Added event from note: ${note.text}")

            scheduleRemoval(note.key)
        }

        private fun scheduleRemoval(key: String) {
            if (shouldDismiss())
                dismissHandler!!.postDelayed(
                    { cancelNotification(key) },
                    dismissDelay.toLong()
                )
        }

        // TODO: Make less calls to shared prefs, cache this.
        private fun shouldDismiss(): Boolean {
            return PreferenceManager.getDefaultSharedPreferences(this)
                .getBoolean(PREF_DISMISS, false)
        }

        companion object {
            var instance: NotificationService? = null
                private set
        }
    }
}

fun extractData(sbn: StatusBarNotification): ShortNotification {
    val innerNote = sbn.notification

    val bigText = innerNote.extras.getCharSequence(Notification.EXTRA_BIG_TEXT)
    val extraText = innerNote.extras.getCharSequence(Notification.EXTRA_TEXT)

    val text = bigText?.toString() ?: (extraText?.toString() ?: "")

    val t = innerNote.extras.getCharSequence(Notification.EXTRA_TITLE)
    val title = t?.toString() ?: ""

    val time = Calendar.getInstance()
    time.timeInMillis = innerNote.`when`

    val user = sbn.user

    println("Attempt at user id: ${user.hashCode()}")

    return ShortNotification(
        id = sbn.id,
        key = sbn.key,
        group = innerNote.group ?: "",
        packageName = sbn.packageName,
        flags = innerNote.flags,
        text = text,
        title = title,
        time = time,
//        user = user
    )
}

private val ALLOWED_PACKAGES = listOf(
    "com.auxbrain.egginc",
    "net.itsjustsomedude.tokens",
    "net.itsjustsomedude.tokens.debug"
)

data class ShortNotification(
    val id: Int,
    val key: String,
    val flags: Int,
    val packageName: String,
    val text: String,
    val title: String,
    val group: String,
    val time: Calendar,
//    val user: Int
) {
    val shouldProcess: Boolean
        get() {
            if (!ALLOWED_PACKAGES.contains(packageName)) return false

            if ((flags and Notification.FLAG_GROUP_SUMMARY) != 0) return false

            if (!title.contains("Gift Received")) return false

            return true
        }
}